modules/hmi_adapter/hmi_stream_adapter_controller.lua

```diff
@@ -0,0 +1,107 @@
+--- Module which provides transport level interface for emulate connection to SDL from HMI for streaming
+--
+-- *Dependencies:* `qt`, `network`
+--
+-- *Globals:* `xmlReporter`, `qt`, `network`
+-- @module tcp_connection
+-- @copyright [SmartDeviceLink Consortium](https://smartdevicelink.com/consortium/)
+-- @license <https://github.com/smartdevicelink/sdl_core/blob/master/LICENSE>
+
+local Stream = { }
+
+--- Type which provides transport level interface for emulate connection with SDL from HMI
+-- @type Connection
+
+--- Construct instance of Connection type
+-- @tparam string host SDL host address
+-- @tparam string port SDL port
+-- @tparam number bytes Number of bytes to receive before calling callback
+-- @tparam function func Callback when stream ends or number of bytes are received
+-- @treturn Connection Constructed instance
+function Stream.TcpConnection(host, port, bytes, func)
+  local res =
+  {
+    host = host,
+    port = port,
+    callback = func,
+    callbackBytes = bytes,
+    receivedBytes = 0,
+    file = io.open("tcp_stream.out","w+b")
+  }
+
+  res.socket = network.TcpClient()
+  res.qtproxy = qt.dynamic()
+
+  function streamTcpCleanup()
+    res.callbackBytes = -1
+    res.file:close()
+    res.file = nil
+  end
+
+  function res.qtproxy.inputData(_, data)
+    if res.receivedBytes == 0 then
+        -- trim off TCP header
+        data = string.sub(data, data:find("\r\n\r\n") + 4)
```

Comment:
Seeing a rare crash here. Was able to reproduce it 2 times out of 9 tries

```
Start './test_scripts/MobileProjection/Phase1/003_Start_audio_streaming.lua'
==============================
SDL build option REMOTE_CONTROL is unavailable.
Assume that SDL was built with REMOTE_CONTROL = ON
Default mobile device transport: TCP
=====================================================================================================
Description: ATF test script
Severity: Major
=====================================================================================================
--- Preconditions -----------------------------------------------------------------------------------
[11:49:53,493] Clean_environment                                                                     [SUCCESS] (405 ms)
SDL pid 26357
tcp        0      0 127.0.0.1:8087          0.0.0.0:*               LISTEN     
SDL started
HMI initialized
HMI is ready
Mobile #1 connected
SDL allowed
[11:49:54,946] Start_SDL__HMI__connect_Mobile__start_Session                                         [SUCCESS] (1484 ms)
[11:49:56,466] Register_App                                                                          [SUCCESS] (2881 ms)
App 1 was used for PTU
[11:49:59,348] PolicyTableUpdate_with_HMI_types                                                      [SUCCESS] (400 ms)
[11:49:59,749] Activate_App                                                                          [SUCCESS] (400 ms)
[11:50:00,150] Start_audio_service                                                                   [SUCCESS] (397 ms)
--- Test --------------------------------------------------------------------------------------------
Streaming...
[11:50:00,548] Start_audio_streaming                                                                 [SUCCESS] (1199 ms)
HMI connected to stream
Lua panic: ./modules/hmi_adapter/hmi_stream_adapter_controller.lua:44: attempt to perform arithmetic on a nil value (/home/jack/Projects/sdl_atf/src/remote_adapter/remote_adapter_client/lua_remote_library.cc:16)
SCRIPT STATUS: ABORTED
'smartDeviceLinkCore' is running, PID: 26357, terminating ...
Done
```

Reply:
This was due to a race condition. Lua had to count the data (usually 65k bytes) before `res.receivedBytes` would be updated. I have changed the way data is saved in https://github.com/smartdevicelink/sdl_atf/pull/234/commits/ee1613e615adf44dc44ac39583c78e51d1bfa8d2

I've run the test about 20 times without a failure